### PR TITLE
implement variable game speed #525

### PR DESCRIPTION
**Pull request by lkk7**

### CODE REVIEW

**Comment by mewmew** [comment]
Really cool patch @lkk7!

I tried running it locally, but it seems that Valgrind found a memory corruption issue. The reason for this is that too little space is allocated for the `speed_str` variable (`char speed_str[2];`).

My `diablo.ini` file contains the following:

```ini
game speed=30
```

As such, at least 3 bytes will be required by this example (to have space for the terminating NULL-byte).

```
=================================================================
==337692==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffcf9a88c42 at pc 0x7f15612854e1 bp 0x7ffcf9a88bd0 sp 0x7ffcf9a88378
READ of size 3 at 0x7ffcf9a88c42 thread T0
    #0 0x7f15612854e0 in __interceptor_atoi /build/gcc/src/gcc/libsanitizer/asan/asan_interceptors.cc:512
    #1 0x7920a0 in run_game_loop Source/diablo.cpp:179
    #2 0x791e05 in StartGame Source/diablo.cpp:117
    #3 0x883639 in mainmenu_init_menu Source/mainmenu.cpp:145
    #4 0x883541 in mainmenu_single_player Source/mainmenu.cpp:133
    #5 0x8833a3 in mainmenu_loop Source/mainmenu.cpp:102
    #6 0x792ec9 in DiabloMain Source/diablo.cpp:309
    #7 0x745cb9 in _cgo_d060868558a9_Cfunc_DiabloMain /tmp/go-build/cgo-gcc-prolog:55
    #8 0x508a9f in runtime.asmcgocall /home/u/go/src/runtime/asm_amd64.s:655

Address 0x7ffcf9a88c42 is located in stack of thread T0 at offset 98 in frame
    #0 0x791ebf in run_game_loop Source/diablo.cpp:154

  This frame has 2 object(s):
    [32, 56) 'msg' (line 157)
    [96, 98) 'speed_str' (line 174) <== Memory access at offset 98 overflows this variable
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow /build/gcc/src/gcc/libsanitizer/asan/asan_interceptors.cc:512 in __interceptor_atoi
Shadow bytes around the buggy address:
```

File: Source/diablo.cpp [hunk 1]
```diff
@@ -168,6 +169,13 @@ void run_game_loop(unsigned int uMsg)
 	force_redraw = 255;
 	gbGameLoopStartup = TRUE;
 	nthread_ignore_mutex(FALSE);
+	char speed_str[2];
```
Comment (mewmew) [inline Source/diablo.cpp]:
```suggestion
	char speed_str[32];

```

File: Source/diablo.cpp [hunk 2]
```diff
@@ -168,6 +169,13 @@ void run_game_loop(unsigned int uMsg)
 	force_redraw = 255;
 	gbGameLoopStartup = TRUE;
 	nthread_ignore_mutex(FALSE);
+	char speed_str[2];
+	if (!getIniValue("devilutionx", "game speed", speed_str, 2)) {
```
Comment (mewmew) [inline Source/diablo.cpp]:
```suggestion
	if (!getIniValue("devilutionx", "game speed", speed_str, sizeof(speed_str))) {

```

File: Source/diablo.cpp [hunk 3]
```diff
@@ -168,6 +169,13 @@ void run_game_loop(unsigned int uMsg)
 	force_redraw = 255;
 	gbGameLoopStartup = TRUE;
 	nthread_ignore_mutex(FALSE);
+	char speed_str[2];
+	if (!getIniValue("devilutionx", "game speed", speed_str, 2)) {
+		game_speed = 50;
+	}
+	else {
+		game_speed = 1000 / std::atoi(speed_str);
```
Comment (mewmew) [inline Source/diablo.cpp]:
```suggestion
				speed_str[sizeof(speed_str)-1] = '\0';
				game_speed = 1000 / std::atoi(speed_str);

```

**Comment by lkk7** [comment]
Thanks for the corrections!

File: Source/diablo.cpp [hunk 4]
```diff
@@ -168,6 +169,17 @@ void run_game_loop(unsigned int uMsg)
 	force_redraw = 255;
 	gbGameLoopStartup = TRUE;
 	nthread_ignore_mutex(FALSE);
+	char speed_str[32];
+
+	if (!getIniValue("devilutionx", "game speed", speed_str, sizeof(speed_str))) {
+
+		game_speed = 50;
+	}
+	else {
```
Comment (mewmew) [inline Source/diablo.cpp]: Minor nit-pick. Put these on the same line.

```diff
-}
-else {
+} else {
```

This is the convention used in the rest of the codebase:

```
$ ack -x cpp "} else {" devilutionX | wc -l
843
```

Comment (qndel) [reply Source/diablo.cpp]: also the empty lines make my eyes bleed and pants wither! :)

**Comment by mewmew** [comment]
@lkk7, it was a lot of fun to play around with this feature :)

To make a proof of concept, I wrote a basic bot which controls the player actions and set `game_speed = 300`. Haha, got a level up within less than 1 minute.

Short video recording of the basic bot in action:
https://cdn.discordapp.com/attachments/540552933700403203/680959190881402900/djavul-ai-0.mp4

**Comment by lkk7** [comment]
Hah, I didn't want to dismiss your review. I changed the file according to your suggestions. It shouldn't show as dismissed! Just resolved.

**Comment by lkk7** [comment]
Hah, I didn't want to dismiss your correction! I changed the style according to your suggestions. It shouldn't be dismissed, just resolved!

**Comment by AJenbo** [comment]
this has been manually merged via https://github.com/diasurgical/devilutionX/commit/f98651d1b8fb4883f6adb3ac04ce64245032578d

**Comment by mewmew** [comment]
> this has been manually merged via f98651d

That's great! I really enjoy this feature and love that it has now been added to DevilX proper.

**Pull request closed without merging**